opt: normalize greatest and least functions into separate comparison clauses

cty123 opened the pull request:
As raised in the issue #119831 it does make sense to normalize the `greatest` and `least` functions such that the index can be leveraged during the query. 

Before the change,

```sql
root@127.0.0.1:26257/defaultdb> EXPLAIN select * from alphabet where least(b) < 30; 

---------------------------------------------------------------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • filter
  │ estimated row count: 0
  │ filter: least(b) < 30
  │
  └── • scan
        estimated row count: 1 (100% of the table; stats collected 16 seconds ago; using stats forecast for 44 seconds in the future)
        table: alphabet@alphabet_pkey
        spans: FULL SCAN
```

after

```sql
root@127.0.0.1:26257/defaultdb> EXPLAIN select * from alphabet where least(b) < 30; 

  • index join
  │ estimated row count: 1
  │ table: alphabet@alphabet_pkey
  │
  └── • scan
        estimated row count: 1 (100% of the table; stats collected 52 minutes ago; using stats forecast for 51 minutes ago)
        table: alphabet@alphabet_b_idx
        spans: (/NULL - /29]
```

Also, the normalization allows further optimizations by other opt rules. For example, 

```sql
SELECT * FROM xxx WHERE greatest(x, 3) < 42
```

is essentially equivalent to 

```sql
SELECT * FROM xxx WHERE x < 42
```

I have added test cases for the OPT rules, and it seems the normalization flow is able to optimize away the redundent clause.

Epic: none

Release note: None

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=119979) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=119979) before we can accept your contribution.<br/><hr/>**Tianyu Chen** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=119979) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119979)
<!-- Reviewable:end -->

mgartner reviewed (changes requested):
Unfortunately, there's some tricky NULL-related behavior that makes this rule invalid as written—see my comment below.

I haven't confirmed, but there might also be some tricky behavior if the types of the `greatest`/`least` arguments do not match the RHS of the comparsion. We should think more about that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119979#-:-NsJMnBk9GoGJSEil5DT:b-vww60y)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @cty123 and @michae2)

___
*[`-- commits` line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119979#-NsJDwQ81fgwq0l0odhc):*
nit: wrap at 100 characters or less, I think 72 characters is most common for our commit messages.
___
*[`pkg/sql/opt/norm/comp_funcs.go` line 165 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119979#-NsJGsrJ3XPCwLDI5Eqh:-NsJGsrJ3XPCwLDI5Eqi:bv8of5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/533f4ee328a67d16131deed67ff28949fc333ede/pkg/sql/opt/norm/comp_funcs.go#L165)):*
> ```Go
> // into:
> //
> //	(x < 42) AND (y < 42)
> ```

Unfortunately, this transformation isn't valid in the presence of NULLs:

```
defaultdb> SELECT greatest(1, NULL) < 42;
  ?column?
------------
     t
(1 row)

defaultdb> SELECT 1 < 42 AND NULL < 42;
  ?column?
------------
    NULL
(1 row)
```

I thought maybe we could do something like this instead:

```
greatest(x, y) < 42
=>
(x IS NULL OR x < 42) AND (y IS NULL OR y < 42)
```

but this is again invalid when all the values are null:

```
defaultdb> SELECT greatest(NULL, NULL) < 42;
  ?column?
------------
  NULL
(1 row)

defaultdb> SELECT (NULL IS NULL OR NULL < 42) AND (NULL IS NULL OR NULL < 42);
  ?column?
------------
     t
(1 row)
```

If we changed the rule to operate on the filters of `Select` relational expressions, we can access logical properties that will tell is if a column is non-NULL by definition or by some other filter lower down in the expression tree. Example: https://github.com/cockroachdb/cockroach/blob/cc4d6574a3f0442ba07c8f1d26d22509bed80d5b/pkg/sql/opt/norm/rules/select.opt#L328.

Then _I think_ we can apply the rule only when we have at least one non-null column or non-null constant.

I'll continue to think about this... There might be a better solution.
___
*[`pkg/sql/opt/norm/comp_funcs.go` line 185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119979#-NsJFXND9rDm4Nn_Jfam:-NsJFXND9rDm4Nn_Jfan:b-7snefv) ([raw file](https://github.com/cockroachdb/cockroach/blob/533f4ee328a67d16131deed67ff28949fc333ede/pkg/sql/opt/norm/comp_funcs.go#L185)):*
> ```Go
> 			node = c.f.ConstructLe(args[i], right)
> 		case opt.LtOp:
> 			node = c.f.ConstructLt(args[i], right)
> ```

Rather than a `switch` you should be able to use `c.f.DynamicConstruct(op, args[i], right)`.


<!-- Sent from Reviewable.io -->

mgartner commented:
Also I rudely forgot to say: Thanks for the contribution!!! And great job figuring out the optgen DSL and custom functions! I think there is likely a way to make this rule valid, with some tweaks, and this is a good starting point.





<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119979)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @cty123 and @mgartner)

___
*[`pkg/sql/opt/norm/comp_funcs.go` line 165 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119979#-NsJGsrJ3XPCwLDI5Eqh:-NsJjrAHC_iFuiC8PVDG:b-nyi3eb) ([raw file](https://github.com/cockroachdb/cockroach/blob/533f4ee328a67d16131deed67ff28949fc333ede/pkg/sql/opt/norm/comp_funcs.go#L165)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Unfortunately, this transformation isn't valid in the presence of NULLs:

```
defaultdb> SELECT greatest(1, NULL) < 42;
  ?column?
------------
     t
(1 row)

defaultdb> SELECT 1 < 42 AND NULL < 42;
  ?column?
------------
    NULL
(1 row)
```

I thought maybe we could do something like this instead:

```
greatest(x, y) < 42
=>
(x IS NULL OR x < 42) AND (y IS NULL OR y < 42)
```

but this is again invalid when all the values are null:

```
defaultdb> SELECT greatest(NULL, NULL) < 42;
  ?column?
------------
  NULL
(1 row)

defaultdb> SELECT (NULL IS NULL OR NULL < 42) AND (NULL IS NULL OR NULL < 42);
  ?column?
------------
     t
(1 row)
```

If we changed the rule to operate on the filters of `Select` relational expressions, we can access logical properties that will tell is if a column is non-NULL by definition or by some other filter lower down in the expression tree. Example: https://github.com/cockroachdb/cockroach/blob/cc4d6574a3f0442ba07c8f1d26d22509bed80d5b/pkg/sql/opt/norm/rules/select.opt#L328.

Then _I think_ we can apply the rule only when we have at least one non-null column or non-null constant.

I'll continue to think about this... There might be a better solution.
</blockquote></details>

Great catch!

Could we add `AND (x IS NOT NULL OR y IS NOT NULL)` to your second idea? I.e.

```
greatest(x, y) < 42
=>
(x IS NULL OR x < 42) AND (y IS NULL OR y < 42) AND (x IS NOT NULL OR y IS NOT NULL)
```


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/norm/comp_funcs.go` line 165 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119979#-NsJGsrJ3XPCwLDI5Eqh:-NsKbBwM0qr2PQ_NaaTP:bivn9gy) ([raw file](https://github.com/cockroachdb/cockroach/blob/533f4ee328a67d16131deed67ff28949fc333ede/pkg/sql/opt/norm/comp_funcs.go#L165)):*
Smart, that's basically codifying this requirement:

> Then _I think_ we can apply the rule only when we have at least one non-null column or non-null constant.

It might work, but only in the context of a filter where `NULL` is falsy, because it's not semantically identical in the general case:

```
defaultdb> SELECT greatest(NULL, NULL) < 42;
  ?column?
------------
  NULL
(1 row)

defaultdb> select (NULL IS NULL OR NULL < 42) AND (NULL IS NULL OR NULL < 42) AND (NULL IS NOT NULL OR NULL IS NOT NULL);
  ?column?
------------
     f
(1 row)
```


<!-- Sent from Reviewable.io -->

cty123 commented:
Thanks for the review, I actually didn't think much about null case. Let me experiment a bit, at least we should be able to optimize when none of those is null.

mgartner commented:
Another open question: Does this transformation add evaluation overhead in the cases where the result expression can't be simplified? In that case, it might not be a good candidate for a normalization rule.

cty123 commented:
I now seem to understand the logic behind this, so similarly, for least function

```sql
SELECT 
    x,
    y,
    least(x, y) < 42 AS "least",
    (
        x IS NOT NULL
        AND x < 42
    )
    OR (
        y IS NOT NULL
        AND y < 42
    )
    OR (
        x IS NULL
        AND y IS NULL
        AND NULL
    ) AS "xform"
FROM (
        VALUES (NULL, NULL),
            (NULL, 1),
            (NULL, 100),
            (1, NULL),
            (100, NULL),
            (1, 2),
            (1, 200),
            (100, 2),
            (100, 200)
    ) v(x, y);
    
-- x   |  y   | least | xform
-- -------+------+-------+--------
--   NULL | NULL | NULL  | NULL
--   NULL |    1 |   t   |   t
--   NULL |  100 |   f   |   f
--      1 | NULL |   t   |   t
--    100 | NULL |   f   |   f
--      1 |    2 |   t   |   t
--      1 |  200 |   t   |   t
--    100 |    2 |   t   |   t
--    100 |  200 |   f   |   f
```

it's a bit contraversal to make this a normalization rule, instead it might be a good one for the cost optimization rule, since it is capable of unpacking the function parameters and leveraging indicies. I am not very familar with memo part, do you know any good place we can put this piece?

cty123 committed:
opt: normalize greatest and least functions into separate comparison clauses

Normalize the `greatest` and `least` functions such that the index can be
leveraged during the query. Also, the normalization allows further optimizations
by other opt rules.

Epic: None

Release note: None

cty123 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119979)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @michae2)

___
*[`pkg/sql/opt/norm/comp_funcs.go` line 185 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119979#-NsJFXND9rDm4Nn_Jfam:-Nsje5YKDVrUZ81qV9of:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/533f4ee328a67d16131deed67ff28949fc333ede/pkg/sql/opt/norm/comp_funcs.go#L185)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Rather than a `switch` you should be able to use `c.f.DynamicConstruct(op, args[i], right)`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner commented:
It may not be all that controversial. As a normalization rule it would certainly make the query plan a bit more confusing to read, e.g., "how did my least(..) filter turn into this?!".

It would be nice to benchmark the two expressions. If our evaluation of the normalized rule is much slower than our evaluation of `least(...) > ...`, then that would rule out the option of making this a normalization rule.

If we only want to perform this in an exploration rule, the first place I'd try that would be to add support in the [`idxconstraint`](https://github.com/cockroachdb/cockroach/tree/master/pkg/sql/opt/idxconstraint) library for building index constraints from expressions like this.
